sklearn/preprocessing/_discretization.py

```diff
@@ -177,6 +183,16 @@ def fit(self, X, y=None):
                 bin_edges[jj] = (centers[1:] + centers[:-1]) * 0.5
                 bin_edges[jj] = np.r_[col_min, bin_edges[jj], col_max]
 
+            # Remove redundant bins (i.e., bins whose width <= 0)
+            if self.strategy in ('quantile', 'kmeans'):
+                mask = np.ediff1d(bin_edges[jj], to_begin=np.inf) > 1e-8
```

Comment:
don't we need to test this for a 32bit system as well? It'd be nice to have a test which tests this precision level and see if it works on all CI machines.

Reply:
I actually think we should make this 0 and not specify a tolerance for a negligible bin. With unary or ordinal encoding small bins won't harm the system... For one hot we could consider a user-specified encoding

Reply:
Sorry, fixed typo

Reply:
I see what you mean, but in general, comparing floats to 0 has kinda proven itself unreliable on multiple occasions at least in the past few months. The last one I remember is the test you commented on, where the two floats logically should have been identical, but they weren't.

`math.isclose()` has the default `rel_tol=1e-9`, which kinda seems like a good value for 64bit systems to me, not sure if it'd be as good on the 32bit ones.

Reply:
> don't we need to test this for a 32bit system as well?

We have 32bit CI?

> I actually think we should make this 0 and not specify a tolerance for a negligible bin.

Is it good to do so when comparing floats? If we use 0, this PR won't solve the new issue completely (see https://github.com/scikit-learn/scikit-learn/pull/13165#discussion_r258322064), or do you want to ignore these extreme cases?

I choose 1e-8 because I saw it several times in the repo and it's also the default atol of np.isclose, not sure if there're better ways.

Reply:
> We have 32bit CI?

we have at least one 32bit windows.

> I choose 1e-8 because I saw it several times in the repo and it's also the default atol of np.isclose, not sure if there're better ways.

AFAIK, `np.isclose` is the way it is for backward compatibility. The new standard is the one adopted by python>=3.5 in [PEP-485](https://www.python.org/dev/peps/pep-0485/) (i.e. `math.isclose`), and seems like a logical choice for new code to follow that one wherever possible/convenient.